t2t/data/dataset_readers/dataset_utils/contrastive_utils.py

```diff
@@ -58,16 +70,49 @@ def sample_anchor_positives(
-    # overlap with each other and the anchor. Their length is sampled from a beta distribution
-    # skewed towards shorter spans.
+    # together is the same document is likely to be semantically similar.
     positives = []
     for _ in range(num_spans):
+        # Their length is sampled from a beta distribution skewed towards shorter spans. The idea
+        # is to promote diversity and minimize the amount of overlapping text.
         positive_length = int(np.random.beta(2, 4) * (max_span_len - min_span_len) + min_span_len)
-        # Be careful not to run off the edges of the document, as this error will pass silently.
-        positive_start = randint(
-            max(0, anchor_start - positive_length), min(anchor_end, num_tokens - positive_length),
-        )
+        # A user can specify a subsuming or adjacent only sampling strategy.
+        if sampling_strategy == "subsuming":
+            positive_start = randint(anchor_start, anchor_end - positive_length)
+        elif sampling_strategy == "adjacent":
+            # We have to restrict positives to a length that will allow them to be adjacent to
+            # the anchor without running off the edge of the document. If documents are sufficiently
+            # long, this won't be a problem and the max_positive_length will equal max_span_len.
+            max_positive_len = min(max_span_len, max(anchor_start, num_tokens - anchor_end))
+            if positive_length > max_positive_len:
+                logger.warning_once(
+                    (
+                        f"There is no room to sample an adjacent positive span of"
+                        f" max_span_len {positive_length}. Temporarily reducing the"
+                        f" maximum span length of positives to {max_positive_len}."
+                        " This message will not be displayed again."
+                    )
+                )
+            positive_length = int(
+                np.random.beta(2, 4) * (max_positive_len - min_span_len) + min_span_len
+            )
+            # There are two types of adjacent positives, those that border the beginning of the
+            # anchor and those that border the end. The checks above guarantee at least one of these
+            # is valid.
+            valid_starts = []
+            if anchor_start - positive_length > 0:
+                valid_starts.append(anchor_start - positive_length)
+            if anchor_end + positive_length <= num_tokens:
+                valid_starts.append(anchor_end)
+            positive_start = choice(valid_starts)
+        else:
+            # By default, spans may be adjacent or overlap with each other and the anchor.
+            # Be careful not to run off the edges of the document, as this error will pass silently.
+            positive_start = randint(
+                max(0, anchor_start - positive_length),
+                min(anchor_end, num_tokens - positive_length),
+            )
+
```

Comment:
Sorry, @OsvaldN mind doing one more quick check?

I realized that throwing an error when there's no room to select an adjacent positive is too aggressive. Instead, I temporarily reduce `max_span_len` to the maximum adjacent span that will fit without going off the edges of the document. We log this as a warning (only logged once so that we don't fill up the console).

Reply:
looks good to me, you're right this is probably a better call than an error.